src/test/java/org/assertj/core/api/InstanceOfAssertFactoriesTest.java

```diff
@@ -0,0 +1,105 @@
+  }
+
+  @Test
+  public void typed_predicate_factory_should_allow_typed_predicate_assertions() {
+    Predicate<String> predicate = Strings::isNullOrEmpty;
+    assertThat((Object) predicate).instanceOf(predicate(String.class)).accepts("");
+  }
+
+  @Test
+  public void int_predicate_factory_should_allow_int_predicate_assertions() {
+    IntPredicate predicate = i -> i == 0;
+    assertThat((Object) predicate).instanceOf(INT_PREDICATE).accepts(0);
+  }
+
+  @Test
+  public void long_predicate_factory_should_allow_long_predicate_assertions() {
+    LongPredicate predicate = l -> l == 0L;
+    assertThat((Object) predicate).instanceOf(LONG_PREDICATE).accepts(0L);
+  }
+
+  @Test
+  public void double_predicate_factory_should_allow_double_predicate_assertions() {
+    DoublePredicate predicate = d -> d == 0.0;
+    assertThat((Object) predicate).instanceOf(DOUBLE_PREDICATE).accepts(0.0);
+  }
+
+  @Test
+  public void completable_future_factory_should_allow_completable_future_assertions() {
+    CompletableFuture<Object> completableFuture = completedFuture("done");
+    assertThat((Object) completableFuture).instanceOf(COMPLETABLE_FUTURE).isDone();
+  }
+
+  @Test
+  public void typed_completable_future_factory_should_allow_typed_completable_future_assertions() {
+    CompletableFuture<String> completableFuture = completedFuture("done");
+    assertThat((Object) completableFuture).instanceOf(completableFuture(String.class)).isDone();
+  }
+
+  @Test
+  public void completion_stage_factory_should_allow_completable_future_assertions() {
+    CompletionStage<Object> completableFuture = completedFuture("done");
+    assertThat((Object) completableFuture).instanceOf(COMPLETION_STAGE).isDone();
+  }
+
+  @Test
+  public void typed_completion_stage_factory_should_allow_typed_completable_future_assertions() {
+    CompletionStage<String> completableFuture = completedFuture("done");
+    assertThat((Object) completableFuture).instanceOf(completionStage(String.class)).isDone();
+  }
+
+  @Test
+  public void string_factory_should_allow_string_assertions() {
+    assertThat((Object) "string").instanceOf(STRING).startsWith("str");
+  }
+
+  @Test
+  public void integer_factory_should_allow_integer_assertions() {
+    assertThat((Object) 0).instanceOf(INTEGER).isZero();
+  }
+
+  @Test
+  @SuppressWarnings("deprecation")
+  public void class_factory_should_allow_class_assertions() {
+    assertThat((Object) Certificate.class).instanceOf(CLASS).hasAnnotations(Deprecated.class);
+  }
+
+  @Test
+  public void list_factory_should_allow_list_assertions() {
+    List<Object> list = asList("Homer", "Marge", "Bart", "Lisa", "Maggie");
+    assertThat((Object) list).instanceOf(LIST).contains("Bart", "Lisa");
+  }
+
+  @Test
+  public void typed_list_factory_should_allow_typed_list_assertions() {
+    List<String> list = asList("Homer", "Marge", "Bart", "Lisa", "Maggie");
+    assertThat((Object) list).instanceOf(list(String.class)).contains("Bart", "Lisa");
```

Comment:
replace the assertion by this one that better shows the benefit of `instanceOf`:
```java
assertThat(list).first().instanceOf(STRING).startsWith("Ho");
```

Reply:
The purpose of this test case is to show the typed List cast, the `String` one is already covered in `string_factory_should_allow_string_assertions()`.

Actually, this test case could be more expressive with:

`assertThat((Object) list).instanceOf(list(String.class)).first().startsWith("Ho");`

but I just noticed `first()` is returning an `ObjectAssert`. I'll investigate a bit if this can easily be changed.

Reply:
I'm happy with the refactored tests you have pushed, as for `first()` let's discuss that in another issue/PR, this one is big enough ;-)

Reply:
I investigated a bit and the main issue for `first()` is the `ObjectAssertFactory` in the `ListAssert` constructor, some refactoring would be needed there. I'll open a separate issue about it.